src/pybind/mgr/cephadm/upgrade.py

```diff
@@ -272,14 +322,103 @@ def upgrade_start(self, image: str, version: str) -> str:
         self.mgr.log.info('Upgrade: Started with target %s' % target_name)
         self.upgrade_state = UpgradeState(
             target_name=target_name,
-            progress_id=str(uuid.uuid4())
+            progress_id=str(uuid.uuid4()),
+            daemon_types=daemon_types,
+            hosts=hosts,
+            services=services,
+            total_count=limit,
+            remaining_count=limit,
         )
         self._update_upgrade_progress(0.0)
         self._save_upgrade_state()
         self._clear_upgrade_health_checks()
         self.mgr.event.set()
         return 'Initiating upgrade to %s' % (target_name)
 
+    def _validate_upgrade_filters(self, target_name: str, daemon_types: Optional[List[str]] = None, hosts: Optional[List[str]] = None, services: Optional[List[str]] = None) -> None:
+        def _latest_type(dtypes: List[str]) -> str:
+            # [::-1] gives the list in reverse
+            for daemon_type in CEPH_UPGRADE_ORDER[::-1]:
+                if daemon_type in dtypes:
+                    return daemon_type
+            return 'N/A'
```

Comment:
is there any special reason for not using `None` in this case instead of `N/A`?

Reply:
no real reason for `'N/A'` in particular. Wanted it to not be `None` so I didn't have to worry about typing and would always get a string back, but honestly it should really just be an empty string by that logic rather than `'N/A'`. I'll make that change,

Reply:
changed this to return an empty string instead